### ISSUE TITLE
Expose IsPreprocessorSymbolDefined

### ISSUE THREAD

**Issue by 333fred** [dotnet/roslyn#24569]
For https://github.com/dotnet/roslyn-analyzers/issues/1241 (and similar analyzers looking for empty method bodies), we need to be able to tell whether the method body in question is conditionally included, so we don't flag methods that only contain `Debug` calls that are wrapped in `#if DEBUG` directives. Currently, we have an API that could provide this information, http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/CSharpExtensions.cs,009faa64ba9f0af9,references, but it's internal. We should look at exposing this API on SemanticModel, possibly swapping as an overload with a `SyntaxNode` argument as the location instead of an `int` for position.

**Comment by gafter** [comment]
The compiler does not parse code that is conditionally included (e.g. `#if DEBUG`) when the preprocessor symbol (e.g. `DEBUG`) is not set. So any invocation found in the parse tree is definitely part of the program.

I'm not saying that there are no use cases for this feature request, but I don't see how the stated use case requires this proposed API.

**Comment by 333fred** [comment]
I admittedly haven't tried to repro this personally, but @mavasani suggested that this is not the case. https://github.com/dotnet/roslyn-analyzers/issues/1241#issuecomment-310740456.

**Comment by Evangelink** [comment]
@gafter, @333fred You are right when you say that the code isn't processed in the symbol is not found but that's not exactly what we need to update CA1821. The rule doesn't allow empty finalizer, all conditional method calls being excluded from the count _EXCEPT WHEN_ the finalizer is itself conditional.

For example this use case should raise a diagnostic:

```csharp
public class C
{
	~C()
	{
		System.Diagnostics.Debug.Fail("something");
	}
}
```

while this use case shouldn't raise a diagnostic:

```csharp
public class C
{
#if DEBUG
	~C()
	{
		System.Diagnostics.Debug.Fail("something");
	}
#endif
}
```

Now, I don't think that the method `IsPreprocessorSymbolDefined` is the one we do need. IMO we would need to either:

1. Access to all the directives available in the context of the current symbol, so we can check if there is at least one invocation/statement present.

2. Be able to merge results from the various compilations before raising any diagnostic.

**Comment by 333fred** [comment]
I'm not sure why `IsPreprocessorSymbolDefined` wouldn't work? You'd look at the conditional attribute on the method, see that it's conditional on `DEBUG`, and ask if `DEBUG` is defined. If it is, then everything is good. If it isn't, then you're not.

**Comment by Evangelink** [comment]
Simply because I am stupid and didn't get that `symbolName` was the name of the directive symbol.
